torch/nn/parallel/data_parallel.py

```diff
@@ -68,8 +143,73 @@ def forward(self, *inputs, **kwargs):
         outputs = self.parallel_apply(replicas, inputs, kwargs)
         return self.gather(outputs, self.output_device)
 
+    def __getstate__(self):
+        state = super(DataParallel, self).__getstate__().copy()
+        del state['_invalidate_cache'], state['_replicas'], state['_orig_params'], \
+            state['_orig_buffers'], state['_repl_params'], state['_repl_buffers']
+        return state
+
+    def __setstate__(self, state):
+        super(DataParallel, self).__setstate__(state)
+        self._init_cache()
+
     def replicate(self, module, device_ids):
-        return replicate(module, device_ids)
+        if not self._replicas:
```

Comment:
So, IIUC, all of these shenanigans are so that we don't break BC by having DataParallel end up with "stale" replicas when someone edits one of the submodules, correct?  Would it be possible to also add a "no magic" version of DataParallel which applies the breaking change but doesn't require all of these shenanigans to detect when to invalidate the cache?

Reply:
I don't think that's a good idea. We've had that in Lua and people kept making mistakes all the time.